### PR TITLE
[Cache] Lock-less FilesystemAdapter

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 3.1 |
| Bug fix? | yes (perf) |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php [hunk 1]
```diff
@@ -138,8 +136,9 @@ protected function doSave(array $values, $lifetime)
                 @mkdir($dir, 0777, true);
             }
             $value = $expiresAt."\n".rawurlencode($id)."\n".serialize($value);
-            if (false !== @file_put_contents($file, $value, LOCK_EX)) {
-                @touch($file, $expiresAt);
+            if (false !== @file_put_contents($tmp, $value)) {
+                @touch($tmp, $expiresAt);
+                $ok = @rename($tmp, $file) && $ok;
```
Comment (phansys) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: IMO, cheaper check should be moved left: `$ok = $ok && @rename($tmp, $file);`

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: It wouldn't call rename all the time, which is required here.

Comment (phansys) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: IIUC, the only reason to keep the current order is to always execute `rename()` regardless the value finally assigned to `$ok`. Is it right?

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: @phansys right

Comment (stof) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: @nicolas-grekas I suggest putting a comment here to make it clear for people reading the code

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: I feel like If I add a comment for this kind of code lines, I should add one on every single line... There are several places in the component (and also in doctrine cache) where this style is in use...

**Comment by nicolas-grekas** [comment]
ping @symfony/deciders votes pendind for 3.1-beta which should/may be released tomorrow

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**